[inductor] Generate indirect_indexing checks even if optimized out

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101602
* __->__ #100895
* #101077

Fixes #100831, fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

peterbell10 committed:
[inductor] Generate indirect_indexing checks even if optimized out

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Lint on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1050,45 +1050,25 @@ def mask_loads(self, mask):
-        if mask == "None":
-            return
-        body = self.current_node._body
-        indirect_size = dict(zip(body.indirect_vars, body.indirect_max_sizes))
```
You can then also revert the code I added that populates ìndirect_max_sizes` and similar structures.

peterbell10 replied:
I spotted a problem with this. Theoretically you can have one maximum under one set of masks, and another under a different set of masks. Not an insurmountable problem, but I suspect it wouldn't be worth the complexity.

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1050,45 +1050,25 @@ def mask_loads(self, mask):
+                )
+                cond = f"({cond}) | ~{mask}"
+            line = f'tl.device_assert({cond}, "index out of bounds: {cond_print}")'
+            self.cse.generate(self.compute, line, assignment=False)
```
Is it alright to always generate this in `self.compute`? Can't the load be in `self.load`, which happens before `self.compute`?

peterbell10 replied:
Indirect loads are always in the `compute` buffer:
https://github.com/pytorch/pytorch/blob/660a0d8622d79fdc1a2b9a7102a5ecd33096c2ec/torch/_inductor/codegen/common.py#L717-L719

lezcano reviewed (commented):
I think that an index variable is not often used in different expressions, so that should be fine.  We should benchmark this though.

I like how much more self-contained the implementation is.

lezcano commented:
The issue with repeated asserts could potentially be solved via some mechanism similar to `DeferredLine`, but rather than removing the line, you could have a way to swap it.

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

One downside of this is that if the same variable is used in two calls to 
`ops.indirect_indexing` with different sizes, we will generate two asserts
instead of one. e.g. before we got
```python
tl.device_assert(((0 <= tmp2) & (tmp2 < 2)) | (~xmask), f"index out of bounds: 0 <= tmp2 < 2")
```
where now it generates two asserts
```python
tl.device_assert((((0 <= tmp2) & (tmp2 < 3))) | ~xmask, "index out of bounds: 0 <= tmp2 < 3")
tl.device_assert((((0 <= tmp2) & (tmp2 < 2))) | ~xmask, "index out of bounds: 0 <= tmp2 < 2")
```

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented:
> The issue with repeated asserts could potentially be solved via some mechanism similar to `DeferredLine`, but rather than removing the line, you could have a way to swap it.

I spotted a problem with this. Theoretically you can have one maximum under one set of masks, and another under a different set of masks. Not an insurmountable problem, but I suspect it wouldn't be worth the complexity.

jgong5 commented:
Instead of assert multiple times, should we reduce the multiple indirect indexing with different sizes into a single assertion with the smallest size?

lezcano commented:
>Instead of assert multiple times, should we reduce the multiple indirect indexing with different sizes into a single assertion with the smallest size?

This is what we were doing before. Also, I think that if we keep the `indirect_max_sizes` list, that should be populated before we generate this assertions, so we should still be able to do that with this approach.

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

One downside of this is that if the same variable is used in two calls to 
`ops.indirect_indexing` with different sizes, we will generate two asserts
instead of one. e.g. before we got
```python
tl.device_assert(((0 <= tmp2) & (tmp2 < 2)) | (~xmask), f"index out of bounds: 0 <= tmp2 < 2")
```
where now it generates two asserts
```python
tl.device_assert((((0 <= tmp2) & (tmp2 < 3))) | ~xmask, "index out of bounds: 0 <= tmp2 < 3")
tl.device_assert((((0 <= tmp2) & (tmp2 < 2))) | ~xmask, "index out of bounds: 0 <= tmp2 < 2")
```

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

One downside of this is that if the same variable is used in two calls to 
`ops.indirect_indexing` with different sizes, we will generate two asserts
instead of one. e.g. before we got
```python
tl.device_assert(((0 <= tmp2) & (tmp2 < 2)) | (~xmask), f"index out of bounds: 0 <= tmp2 < 2")
```
where now it generates two asserts
```python
tl.device_assert((((0 <= tmp2) & (tmp2 < 3))) | ~xmask, "index out of bounds: 0 <= tmp2 < 3")
tl.device_assert((((0 <= tmp2) & (tmp2 < 2))) | ~xmask, "index out of bounds: 0 <= tmp2 < 2")
```

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1050,45 +1052,59 @@ def mask_loads(self, mask):
+                line = 'tl.device_assert({cond}, "index out of bounds: {cond_print}")'
+                size_map[map_key] = size
+                self.compute.writeline(
+                    IndirectAssertLine(line, var_str, mask, size_map)
```
Okay, here is a completely different strategy for de-duplicating the assert lines.  I delay the final formatting of `cond` and `cond_print` until the entire IR has has been executed, and accumulate size information in `TritonKernel.indirect_max_sizes`. This mapping is keyed by variable name and by mask, so it correctly handles masked loads as well.

This also improves upon the current `main` branch which sometimes generates multiple asserts for the same variable. For example, this python code 
```python
import torch

def f(x, y, ind):
    return x[ind] + y[ind]

x = torch.randn(4, 3, device="cuda")
y = torch.randn(2, 3, device="cuda")
ind = torch.zeros(2, 2, dtype=torch.int64, device="cuda")
cfn = torch.compile(f, dynamic=True)
cfn(x, y, ind)
```

Generates two triton asserts on main,
```python
    tmp0 = tl.load(in_ptr0 + (x1), xmask)
    tl.device_assert((((0 <= tmp0) & (tmp0 < 4))) | ~xmask, "index out of bounds: 0 <= tmp0 < 4")
    tmp1 = tl.load(in_ptr1 + (x0 + (3*tmp0)), xmask)
    tl.device_assert((((0 <= tmp0) & (tmp0 < 2))) | ~xmask, "index out of bounds: 0 <= tmp0 < 2")
    tmp2 = tl.load(in_ptr2 + (x0 + (3*tmp0)), xmask)
```


But with this version it generates only one:
```python
    tl.device_assert(((0 <= tmp0) & (tmp0 < 2)) | ~xmask, "index out of bounds: 0 <= tmp0 < 2")
    tmp1 = tl.load(in_ptr1 + (x0 + (3*tmp0)), xmask)
    tmp2 = tl.load(in_ptr2 + (x0 + (3*tmp0)), xmask)
```
So it doesn't work when we generate the I don't really know how much value this is adding.

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly insert
an `IndirectAssertLine` which will be the one and only assert 
generated for a particular combination of index and mask variables.



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1050,45 +1056,70 @@ def mask_loads(self, mask):
+                    else f"({' & '.join(str(v) for v in mask_vars)})"
+                )
+
+            # tl.device_assert doesn't work for constexpr values, and we can't
+            # tell from here if a var is constexpr or not, so promote everything
+            var_str = str(
+                self.cse.generate(
+                    self.compute, f"triton_helpers.promote_to_tensor({var})"
+                )
+            )
```
This is not going to do as per https://github.com/pytorch/pytorch/pull/100880#issuecomment-1540494933. Now, I think that most the "indirect indexing via scalars" (aka direct indexing) we do is not tagged as indirect_indexing, but I guess that there are some times where this is not the case and we do things like:
```python
tmp0 = 0
tl.load(in_ptr0 + tmp0, ...)
```
I don't know when is it that we generate this sort of code though.

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly insert
an `IndirectAssertLine` which will be the one and only assert 
generated for a particular combination of index and mask variables.



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly insert
an `IndirectAssertLine` which will be the one and only assert 
generated for a particular combination of index and mask variables.



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly insert
an `IndirectAssertLine` which will be the one and only assert 
generated for a particular combination of index and mask variables.



cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

vfdev-5 commented on torch/_inductor/codegen/common.py:
```diff
@@ -743,7 +742,8 @@ def inner(*args, **kwargs):
             def indirect_indexing(index_var, size):
-                return sympy_symbol(str(index_var))
+                # Skip CSE since this doesn't return an expression
+                return self.indirect_indexing(index_var, size)
```
Peter, where comes `self` here ?

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Fixes #100831, fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
lint on "[inductor] Generate indirect_indexing checks even if optimized out"


Fixes #100831, fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Fixes #100831, fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

lezcano reviewed (approved):
Great stuff!

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1050,45 +1056,70 @@ def mask_loads(self, mask):
-                var_mask = f"({mask})" if "&" in mask else mask
-                var_mask = f" | ~{var_mask}"
+    def indirect_indexing(self, var, size):
+        class IndirectAssertLine(DeferredLineBase):
```
nvm, I see you're already doing this.

peterbell10 replied:
That's already handled below:
```
        if config.triton.assert_indirect_indexing and torch.version.hip is None:
```

peterbell10 commented on torch/_inductor/triton_helpers.py:
```diff
@@ -4,11 +4,15 @@
 
+@triton.jit
+def promote_to_tensor(x):
+    # Addition promotes to tensor for us
+    return x + tl.zeros((1,), tl.int1)
```
Doesn't seem like either of these work
```
>>> tl.zeros((), tl.int1)
TypeError('0d block_type is forbidden')

>>> tl.broadcast(1.0, tl.zeros((1,), tl.int1)
AttributeError("'constexpr' object has no attribute 'type'")
```

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Fixes #100831, fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Generate indirect_indexing checks even if optimized out"


Fixes #100831, fixes #100878

Previously `gen_assert_indirect_indexing` was only called on the index
expressions passed to `ops.load` and `ops.store` which means if the
variable is optimized out during lowering, we never generate the
assert. This instead makes `ops.indirect_indexing` eagerly generate
the assert statement, whether or not it will be used.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge

ngimel commented:
Can we delay merging this until the next PR is ready to merge (is it?) so that we don't have to live with the perf regression for long?
